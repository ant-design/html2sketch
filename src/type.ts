import SketchFormat from '@sketch-hq/sketch-file-format-ts';
import { SymbolMaster } from './model';
import { ResizingConstraint } from './utils/layout';
import {
  TextHorizontalAlign,
  TextVerticalAlign,
} from './model/Style/TextStyle';
import Color, { ColorParam } from 'html2sketch/model/Style/Color';

// *** 统一的类型定义收口文件 *** //
export * from './model/type';

/**
 * 获得 CG 点的配置
 */
export interface CGPoint {
  x: number;
  y: number;
}

/**
 * group 或 symbol 的 layout 参数类型
 */
export type GroupLayoutType =
  | 'BOTTOM_TO_TOP'
  | 'HORIZONTALLY_CENTER'
  | 'VERTICALLY_CENTER'
  | 'RIGHT_TO_LEFT'
  | 'LEFT_TO_RIGHT'
  | 'TOP_TO_BOTTOM'
  | 'NONE';

/**
 * 对 symbol 进行自定义处理的方法
 */
export type HandleSymbolFn = (symbol: SymbolMaster) => void;

export interface DefaultSymbolParams {
  symbolName?: string;
  /**
   * symbol 本身的 layout
   */
  symbolLayout?: GroupLayoutType;
  /**
   * symbol 内部的图层配置项
   */
  layerParams: SymbolAdjustParams[];
  /**
   * 图层选择器
   */
  selector?: LayerSelector;
}

export enum LayerSelectorMatchMethod {
  Equal,
  Included,
}

/**
 * 选择图层的方法
 */
export interface LayerSelector {
  type: 'classname' | 'class' | 'name' | 'text' | 'tag';
  value: string;
  match?: LayerSelectorMatchMethod;
}

/**
 * 对 symbol 的相关配置项调整的参数
 */
export interface SymbolAdjustParams {
  /**
   * 选择器
   */
  selector: LayerSelector;
  /**
   * sketch 的调整参数
   */
  resizing?: ResizingConstraint[];
  /**
   * symbol 内 group 的布局类型
   */
  layout?: GroupLayoutType;
  /**
   * symbol 内文本的配置参数
   */
  text?: TextParam;
}

/**
 * 文本对象的配置项
 */
export interface TextParam {
  textBehaviour?: SketchFormat.TextBehaviour;
  verticalAlign?: TextVerticalAlign;
  horizontalAlign?: TextHorizontalAlign;
}

/**
 * nodeToSketchSymbol 配置项
 */
export interface NodeToSketchSymbolOptions {
  /**
   * symbol 自己的 layout 类型
   */
  symbolLayout?: GroupLayoutType;
  /**
   * 如果需要对 symbol 进行调整处理
   * 传入这个方法
   */
  handleSymbol?: HandleSymbolFn;
  /**
   * symbol 内部图层的配置项
   */
  layerParams?: SymbolAdjustParams[];
}

/**
 *  =====渐变定义=====
 */

/**
 * 传入颜色间隔点的参数
 */
export interface ColorStopParam {
  color: ColorParam;
  offset: number;
}

export type StopParam = ColorParam | ColorStopParam;

export interface GradientProps {
  /**
   * 渐变类型
   */
  type?: SketchFormat.GradientType;
  /**
   * 终点
   */
  to?: CGPoint;
  /**
   * 起点
   */
  from?: CGPoint;
  /**
   * 色彩间隔点
   */
  stops?: StopParam[];
  /**
   * 渐变名称
   */
  name?: string;
  /**
   * 如果是圆形渐变 则需要传入 radius 值
   */
  radius?: number;
}

export interface ColorStop {
  /**
   * 偏移位置 取值 0-1
   */
  offset?: number;
  /**
   * 颜色对象
   */
  color: Color;
}
