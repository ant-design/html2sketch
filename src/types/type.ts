/**
 * 获得 CG 点的配置
 */
export interface CGPoint {
  x: number;
  y: number;
}

export declare enum LayerClassValue {
  Artboard = 'artboard',
  Bitmap = 'bitmap',
  Border = 'border',
  BorderOptions = 'borderOptions',
  Color = 'color',
  Fill = 'fill',
  Gradient = 'gradient',
  Group = 'group',
  InnerShadow = 'innerShadow',
  Oval = 'oval',
  Page = 'page',
  Polygon = 'polygon',
  Rect = 'rect',
  Rectangle = 'rectangle',
  Shadow = 'shadow',
  ShapeGroup = 'shapeGroup',
  ShapePath = 'shapePath',
  Slice = 'slice',
  Star = 'star',
  SymbolInstance = 'symbolInstance',
  SymbolMaster = 'symbolMaster',
  Text = 'text',
  Triangle = 'triangle',
  Svg = 'svg',
}

export type LayerClassType =
  | 'artboard'
  | 'bitmap'
  | 'border'
  | 'borderOptions'
  | 'color'
  | 'fill'
  | 'gradient'
  | 'group'
  | 'innerShadow'
  | 'ellipse'
  | 'oval' // sketch 中 用的是 oval
  | 'page'
  | 'polygon'
  | 'rect'
  | 'rectangle'
  | 'shadow'
  | 'shapeGroup'
  | 'shapePath'
  | 'slice'
  | 'star'
  | 'symbolInstance'
  | 'symbolMaster'
  | 'text'
  | 'triangle'
  | 'svg';
