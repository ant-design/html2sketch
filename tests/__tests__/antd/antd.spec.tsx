import React from 'react';
import { Button, Modal, Radio } from 'antd';

import { nodeToGroup, nodeToSymbol } from 'html2sketch';
import { isUpdate, render } from '@test-utils';
import {
  saveJSONData,
  setupAntdTestEnv,
  sleep,
  svgIconJSON,
  radioJSON,
  svgButtonJSON,
  defaultModalJSON,
} from './utils';
import { PlusOutlined, UpCircleOutlined } from '@ant-design/icons';
import type SketchFormat from '@sketch-hq/sketch-file-format-ts';

describe('antd 组件库可正常解析', () => {
  it('Radio 单选器', async () => {
    await setupAntdTestEnv();
    render(<Radio checked>html2sketch</Radio>);

    const node = document.getElementById('container') as HTMLDivElement;

    const group = (await nodeToGroup(node)).toSketchJSON();

    if (isUpdate) {
      saveJSONData(group, 'radio');
    }
    const { frame, ...target } = group;
    const { frame: originFrame, ...origin } = radioJSON;

    expect(JSON.parse(JSON.stringify(target))).toEqual(origin);
    expect(Math.round(frame.width)).toEqual(Math.round(originFrame.width));
  });

  describe('Svg', () => {
    it('svg icon', async () => {
      await setupAntdTestEnv();
      render(<PlusOutlined />);

      const node = document.getElementById('container') as HTMLDivElement;

      const group = (await nodeToGroup(node)).toSketchJSON();

      if (isUpdate) {
        saveJSONData(group, 'svg-icon');
      }
      expect(group).toEqual(svgIconJSON);
    });
    it('SVG 和按钮', async () => {
      await setupAntdTestEnv();
      render(
        <Button id="button" icon={<UpCircleOutlined />} type="primary">
          文本
        </Button>,
      );

      const node = document.getElementById('container') as HTMLDivElement;

      const group = (await nodeToGroup(node)).toSketchJSON();

      if (isUpdate) {
        saveJSONData(group, 'svg-button');
      }
      expect(group).toEqual(svgButtonJSON);
    });
  });

  it('Modal', async () => {
    await setupAntdTestEnv();
    render(
      <div style={{ position: 'relative', minHeight: 400 }}>
        <Modal
          visible
          title="Modal"
          mask={false}
          wrapProps={{ id: 'modal' }}
          // centered
          maskClosable
          getContainer={false}
        >
          Content
        </Modal>
      </div>,
    );

    // 需要等 2 秒让 Modal 正常跳出来
    await sleep(2000);

    const node = document.getElementsByClassName(
      'ant-modal',
    )[0] as HTMLDivElement;

    const symbol = (await nodeToSymbol(node)).toSketchJSON();

    if (isUpdate) {
      saveJSONData(symbol, 'default-modal');
    }
    const { frame, ...target } = symbol;
    const { frame: originFrame, ...origin } = defaultModalJSON;

    expect(target.name).toEqual(origin.name);
    expect(Math.round(frame.width)).toEqual(Math.round(originFrame.width));

    expect(symbol._class).toBe('symbolMaster');
    expect(symbol.groupLayout).toStrictEqual({
      _class: 'MSImmutableFreeformGroupLayout',
    });
    expect(symbol.name).toBe('Modal');
    expect(symbol.layers.length).toBe(5);

    const header = symbol.layers[1] as SketchFormat.Group;
    expect(header.groupLayout).toStrictEqual({
      _class: 'MSImmutableInferredGroupLayout',
      axis: 0,
      layoutAnchor: 0,
    });
  });
});
