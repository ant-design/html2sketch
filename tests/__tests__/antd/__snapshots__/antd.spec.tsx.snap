// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`antd 组件库可正常解析 Tooltip 1`] = `
Object {
  "_class": "group",
  "booleanOperation": -1,
  "clippingMaskMode": 0,
  "do_objectID": "UUID",
  "exportOptions": Object {
    "_class": "exportOptions",
    "exportFormats": Array [],
    "includedLayerIds": Array [],
    "layerOptions": 0,
    "shouldTrim": false,
  },
  "frame": Object {
    "_class": "rect",
    "constrainProportions": false,
    "height": 34,
    "width": 40.546875,
    "x": 8,
    "y": 8,
  },
  "groupLayout": Object {
    "_class": "MSImmutableFreeformGroupLayout",
  },
  "hasClickThrough": false,
  "hasClippingMask": false,
  "isFixedToViewport": false,
  "isFlippedHorizontal": false,
  "isFlippedVertical": false,
  "isLocked": false,
  "isTemplate": false,
  "isVisible": true,
  "layerListExpandedType": 0,
  "layers": Array [
    Object {
      "_class": "group",
      "booleanOperation": -1,
      "clippingMaskMode": 0,
      "do_objectID": "UUID",
      "exportOptions": Object {
        "_class": "exportOptions",
        "exportFormats": Array [],
        "includedLayerIds": Array [],
        "layerOptions": 0,
        "shouldTrim": false,
      },
      "frame": Object {
        "_class": "rect",
        "constrainProportions": false,
        "height": 34,
        "width": 40.546875,
        "x": 0,
        "y": 0,
      },
      "groupLayout": Object {
        "_class": "MSImmutableFreeformGroupLayout",
      },
      "hasClickThrough": false,
      "hasClippingMask": false,
      "isFixedToViewport": false,
      "isFlippedHorizontal": false,
      "isFlippedVertical": false,
      "isLocked": false,
      "isTemplate": false,
      "isVisible": true,
      "layerListExpandedType": 0,
      "layers": Array [
        Object {
          "_class": "rectangle",
          "booleanOperation": -1,
          "clippingMaskMode": 0,
          "do_objectID": "UUID",
          "edited": false,
          "exportOptions": Object {
            "_class": "exportOptions",
            "exportFormats": Array [],
            "includedLayerIds": Array [],
            "layerOptions": 0,
            "shouldTrim": false,
          },
          "fixedRadius": 0,
          "frame": Object {
            "_class": "rect",
            "constrainProportions": false,
            "height": 34,
            "width": 40.546875,
            "x": 0,
            "y": 0,
          },
          "hasClippingMask": false,
          "hasConvertedToNewRoundCorners": true,
          "isClosed": true,
          "isFixedToViewport": false,
          "isFlippedHorizontal": false,
          "isFlippedVertical": false,
          "isLocked": false,
          "isTemplate": false,
          "isVisible": true,
          "layerListExpandedType": 0,
          "name": "ant-tooltip-inner",
          "nameIsFixed": false,
          "needsConvertionToNewRoundCorners": false,
          "pointRadiusBehaviour": 1,
          "points": Array [
            Object {
              "_class": "curvePoint",
              "cornerRadius": 6,
              "cornerStyle": 0,
              "curveFrom": "{0, 0}",
              "curveMode": 1,
              "curveTo": "{0, 0}",
              "hasCurveFrom": false,
              "hasCurveTo": false,
              "point": "{0, 0}",
            },
            Object {
              "_class": "curvePoint",
              "cornerRadius": 6,
              "cornerStyle": 0,
              "curveFrom": "{1, 0}",
              "curveMode": 1,
              "curveTo": "{1, 0}",
              "hasCurveFrom": false,
              "hasCurveTo": false,
              "point": "{1, 0}",
            },
            Object {
              "_class": "curvePoint",
              "cornerRadius": 6,
              "cornerStyle": 0,
              "curveFrom": "{1, 1}",
              "curveMode": 1,
              "curveTo": "{1, 1}",
              "hasCurveFrom": false,
              "hasCurveTo": false,
              "point": "{1, 1}",
            },
            Object {
              "_class": "curvePoint",
              "cornerRadius": 6,
              "cornerStyle": 0,
              "curveFrom": "{0, 1}",
              "curveMode": 1,
              "curveTo": "{0, 1}",
              "hasCurveFrom": false,
              "hasCurveTo": false,
              "point": "{0, 1}",
            },
          ],
          "resizingConstraint": 27,
          "resizingType": 0,
          "rotation": 0,
          "shouldBreakMaskChain": false,
          "style": Object {
            "_class": "style",
            "borderOptions": Object {
              "_class": "borderOptions",
              "dashPattern": Array [],
              "isEnabled": true,
              "lineCapStyle": 0,
              "lineJoinStyle": 0,
            },
            "borders": Array [
              Object {
                "_class": "border",
                "color": Object {
                  "_class": "color",
                  "alpha": 1,
                  "blue": 1,
                  "green": 1,
                  "red": 1,
                },
                "contextSettings": Object {
                  "_class": "graphicsContextSettings",
                  "blendMode": 0,
                  "opacity": 1,
                },
                "fillType": 0,
                "gradient": Object {
                  "_class": "gradient",
                  "elipseLength": 1,
                  "from": "{0.5, 0}",
                  "gradientType": 0,
                  "stops": Array [],
                  "to": "{0.5, 1}",
                },
                "isEnabled": true,
                "position": 1,
                "thickness": 0,
              },
            ],
            "colorControls": Object {
              "_class": "colorControls",
              "brightness": 0,
              "contrast": 1,
              "hue": 0,
              "isEnabled": false,
              "saturation": 1,
            },
            "contextSettings": Object {
              "_class": "graphicsContextSettings",
              "blendMode": 0,
              "opacity": 1,
            },
            "do_objectID": "UUID",
            "endMarkerType": 0,
            "fills": Array [
              Object {
                "_class": "fill",
                "color": Object {
                  "_class": "color",
                  "alpha": 0.85,
                  "blue": 0,
                  "green": 0,
                  "red": 0,
                },
                "contextSettings": Object {
                  "_class": "graphicsContextSettings",
                  "blendMode": 0,
                  "opacity": 1,
                },
                "fillType": 0,
                "gradient": Object {
                  "_class": "gradient",
                  "elipseLength": 1,
                  "from": "{0.5, 0}",
                  "gradientType": 0,
                  "stops": Array [],
                  "to": "{0.5, 1}",
                },
                "isEnabled": true,
                "noiseIndex": 0,
                "noiseIntensity": 0,
                "patternFillType": 1,
                "patternTileScale": 1,
              },
            ],
            "innerShadows": Array [],
            "miterLimit": 10,
            "shadows": Array [
              Object {
                "_class": "shadow",
                "blurRadius": 16,
                "color": Object {
                  "_class": "color",
                  "alpha": 0.08,
                  "blue": 0,
                  "green": 0,
                  "red": 0,
                },
                "contextSettings": Object {
                  "_class": "graphicsContextSettings",
                  "blendMode": 0,
                  "opacity": 1,
                },
                "isEnabled": true,
                "offsetX": 0,
                "offsetY": 6,
                "spread": 0,
              },
              Object {
                "_class": "shadow",
                "blurRadius": 6,
                "color": Object {
                  "_class": "color",
                  "alpha": 0.12,
                  "blue": 0,
                  "green": 0,
                  "red": 0,
                },
                "contextSettings": Object {
                  "_class": "graphicsContextSettings",
                  "blendMode": 0,
                  "opacity": 1,
                },
                "isEnabled": true,
                "offsetX": 0,
                "offsetY": 3,
                "spread": -4,
              },
              Object {
                "_class": "shadow",
                "blurRadius": 28,
                "color": Object {
                  "_class": "color",
                  "alpha": 0.05,
                  "blue": 0,
                  "green": 0,
                  "red": 0,
                },
                "contextSettings": Object {
                  "_class": "graphicsContextSettings",
                  "blendMode": 0,
                  "opacity": 1,
                },
                "isEnabled": true,
                "offsetX": 0,
                "offsetY": 9,
                "spread": 8,
              },
            ],
            "startMarkerType": 0,
            "windingRule": 1,
          },
        },
        Object {
          "_class": "text",
          "attributedString": Object {
            "_class": "attributedString",
            "attributes": Array [
              Object {
                "_class": "stringAttribute",
                "attributes": Object {
                  "MSAttributedStringColorAttribute": Object {
                    "_class": "color",
                    "alpha": 1,
                    "blue": 1,
                    "green": 1,
                    "red": 1,
                  },
                  "MSAttributedStringFontAttribute": Object {
                    "_class": "fontDescriptor",
                    "attributes": Object {
                      "name": "PingFangSC-Regular",
                      "size": 14,
                    },
                  },
                  "MSAttributedStringTextTransformAttribute": 0,
                  "kerning": 0,
                  "paragraphStyle": Object {
                    "_class": "paragraphStyle",
                    "alignment": 0,
                    "maximumLineHeight": 22,
                    "minimumLineHeight": 22,
                  },
                  "strikethroughStyle": 0,
                  "underlineStyle": 0,
                },
                "length": 4,
                "location": 0,
              },
            ],
            "string": "text",
          },
          "automaticallyDrawOnUnderlyingPath": false,
          "booleanOperation": -1,
          "clippingMaskMode": 0,
          "do_objectID": "UUID",
          "dontSynchroniseWithSymbol": false,
          "exportOptions": Object {
            "_class": "exportOptions",
            "exportFormats": Array [],
            "includedLayerIds": Array [],
            "layerOptions": 0,
            "shouldTrim": false,
          },
          "frame": Object {
            "_class": "rect",
            "constrainProportions": false,
            "height": 22,
            "width": 24.546875,
            "x": 8,
            "y": 6,
          },
          "glyphBounds": "",
          "hasClippingMask": false,
          "isFixedToViewport": false,
          "isFlippedHorizontal": false,
          "isFlippedVertical": false,
          "isLocked": false,
          "isTemplate": false,
          "isVisible": true,
          "layerListExpandedType": 0,
          "lineSpacingBehaviour": 2,
          "name": "text",
          "nameIsFixed": false,
          "resizingConstraint": 27,
          "resizingType": 0,
          "rotation": 0,
          "shouldBreakMaskChain": false,
          "style": Object {
            "_class": "style",
            "borderOptions": Object {
              "_class": "borderOptions",
              "dashPattern": Array [],
              "isEnabled": true,
              "lineCapStyle": 0,
              "lineJoinStyle": 0,
            },
            "borders": Array [],
            "colorControls": Object {
              "_class": "colorControls",
              "brightness": 0,
              "contrast": 1,
              "hue": 0,
              "isEnabled": false,
              "saturation": 1,
            },
            "contextSettings": Object {
              "_class": "graphicsContextSettings",
              "blendMode": 0,
              "opacity": 1,
            },
            "do_objectID": "UUID",
            "endMarkerType": 0,
            "fills": Array [],
            "innerShadows": Array [],
            "miterLimit": 10,
            "shadows": Array [],
            "startMarkerType": 0,
            "textStyle": Object {
              "_class": "textStyle",
              "encodedAttributes": Object {
                "MSAttributedStringColorAttribute": Object {
                  "_class": "color",
                  "alpha": 1,
                  "blue": 1,
                  "green": 1,
                  "red": 1,
                },
                "MSAttributedStringFontAttribute": Object {
                  "_class": "fontDescriptor",
                  "attributes": Object {
                    "name": "PingFangSC-Regular",
                    "size": 14,
                  },
                },
                "MSAttributedStringTextTransformAttribute": 0,
                "kerning": 0,
                "paragraphStyle": Object {
                  "_class": "paragraphStyle",
                  "alignment": 0,
                  "maximumLineHeight": 22,
                  "minimumLineHeight": 22,
                },
                "strikethroughStyle": 0,
                "underlineStyle": 0,
              },
              "verticalAlignment": 0,
            },
            "windingRule": 1,
          },
          "textBehaviour": 0,
        },
      ],
      "name": "编组",
      "nameIsFixed": false,
      "resizingConstraint": 27,
      "resizingType": 0,
      "rotation": 0,
      "shouldBreakMaskChain": false,
      "style": Object {
        "_class": "style",
        "borderOptions": Object {
          "_class": "borderOptions",
          "dashPattern": Array [],
          "isEnabled": true,
          "lineCapStyle": 0,
          "lineJoinStyle": 0,
        },
        "borders": Array [],
        "colorControls": Object {
          "_class": "colorControls",
          "brightness": 0,
          "contrast": 1,
          "hue": 0,
          "isEnabled": false,
          "saturation": 1,
        },
        "contextSettings": Object {
          "_class": "graphicsContextSettings",
          "blendMode": 0,
          "opacity": 1,
        },
        "do_objectID": "UUID",
        "endMarkerType": 0,
        "fills": Array [],
        "innerShadows": Array [],
        "miterLimit": 10,
        "shadows": Array [],
        "startMarkerType": 0,
        "windingRule": 1,
      },
    },
    Object {
      "_class": "group",
      "booleanOperation": -1,
      "clippingMaskMode": 0,
      "do_objectID": "UUID",
      "exportOptions": Object {
        "_class": "exportOptions",
        "exportFormats": Array [],
        "includedLayerIds": Array [],
        "layerOptions": 0,
        "shouldTrim": false,
      },
      "frame": Object {
        "_class": "rect",
        "constrainProportions": false,
        "height": 32,
        "width": 32,
        "x": 4.265625,
        "y": 34,
      },
      "groupLayout": Object {
        "_class": "MSImmutableFreeformGroupLayout",
      },
      "hasClickThrough": false,
      "hasClippingMask": false,
      "isFixedToViewport": false,
      "isFlippedHorizontal": false,
      "isFlippedVertical": false,
      "isLocked": false,
      "isTemplate": false,
      "isVisible": true,
      "layerListExpandedType": 0,
      "layers": Array [
        Object {
          "_class": "rectangle",
          "booleanOperation": -1,
          "clippingMaskMode": 0,
          "do_objectID": "UUID",
          "edited": false,
          "exportOptions": Object {
            "_class": "exportOptions",
            "exportFormats": Array [],
            "includedLayerIds": Array [],
            "layerOptions": 0,
            "shouldTrim": false,
          },
          "fixedRadius": 0,
          "frame": Object {
            "_class": "rect",
            "constrainProportions": false,
            "height": 32,
            "width": 32,
            "x": 0,
            "y": 0,
          },
          "hasClippingMask": true,
          "hasConvertedToNewRoundCorners": true,
          "isClosed": true,
          "isFixedToViewport": false,
          "isFlippedHorizontal": false,
          "isFlippedVertical": false,
          "isLocked": false,
          "isTemplate": false,
          "isVisible": true,
          "layerListExpandedType": 0,
          "name": "ant-tooltip-arrow",
          "nameIsFixed": false,
          "needsConvertionToNewRoundCorners": false,
          "pointRadiusBehaviour": 1,
          "points": Array [
            Object {
              "_class": "curvePoint",
              "cornerRadius": 0,
              "cornerStyle": 0,
              "curveFrom": "{0, 0}",
              "curveMode": 1,
              "curveTo": "{0, 0}",
              "hasCurveFrom": false,
              "hasCurveTo": false,
              "point": "{0, 0}",
            },
            Object {
              "_class": "curvePoint",
              "cornerRadius": 0,
              "cornerStyle": 0,
              "curveFrom": "{1, 0}",
              "curveMode": 1,
              "curveTo": "{1, 0}",
              "hasCurveFrom": false,
              "hasCurveTo": false,
              "point": "{1, 0}",
            },
            Object {
              "_class": "curvePoint",
              "cornerRadius": 2,
              "cornerStyle": 0,
              "curveFrom": "{1, 1}",
              "curveMode": 1,
              "curveTo": "{1, 1}",
              "hasCurveFrom": false,
              "hasCurveTo": false,
              "point": "{1, 1}",
            },
            Object {
              "_class": "curvePoint",
              "cornerRadius": 0,
              "cornerStyle": 0,
              "curveFrom": "{0, 1}",
              "curveMode": 1,
              "curveTo": "{0, 1}",
              "hasCurveFrom": false,
              "hasCurveTo": false,
              "point": "{0, 1}",
            },
          ],
          "resizingConstraint": 27,
          "resizingType": 0,
          "rotation": 0,
          "shouldBreakMaskChain": false,
          "style": Object {
            "_class": "style",
            "borderOptions": Object {
              "_class": "borderOptions",
              "dashPattern": Array [],
              "isEnabled": true,
              "lineCapStyle": 0,
              "lineJoinStyle": 0,
            },
            "borders": Array [
              Object {
                "_class": "border",
                "color": Object {
                  "_class": "color",
                  "alpha": 0.88,
                  "blue": 0,
                  "green": 0,
                  "red": 0,
                },
                "contextSettings": Object {
                  "_class": "graphicsContextSettings",
                  "blendMode": 0,
                  "opacity": 1,
                },
                "fillType": 0,
                "gradient": Object {
                  "_class": "gradient",
                  "elipseLength": 1,
                  "from": "{0.5, 0}",
                  "gradientType": 0,
                  "stops": Array [],
                  "to": "{0.5, 1}",
                },
                "isEnabled": true,
                "position": 1,
                "thickness": 0,
              },
            ],
            "colorControls": Object {
              "_class": "colorControls",
              "brightness": 0,
              "contrast": 1,
              "hue": 0,
              "isEnabled": false,
              "saturation": 1,
            },
            "contextSettings": Object {
              "_class": "graphicsContextSettings",
              "blendMode": 0,
              "opacity": 1,
            },
            "do_objectID": "UUID",
            "endMarkerType": 0,
            "fills": Array [],
            "innerShadows": Array [],
            "miterLimit": 10,
            "shadows": Array [],
            "startMarkerType": 0,
            "windingRule": 1,
          },
        },
        Object {
          "_class": "rectangle",
          "booleanOperation": -1,
          "clippingMaskMode": 0,
          "do_objectID": "UUID",
          "edited": false,
          "exportOptions": Object {
            "_class": "exportOptions",
            "exportFormats": Array [],
            "includedLayerIds": Array [],
            "layerOptions": 0,
            "shouldTrim": false,
          },
          "fixedRadius": 0,
          "frame": Object {
            "_class": "rect",
            "constrainProportions": false,
            "height": 8,
            "width": 32,
            "x": 0,
            "y": 24,
          },
          "hasClippingMask": false,
          "hasConvertedToNewRoundCorners": true,
          "isClosed": true,
          "isFixedToViewport": false,
          "isFlippedHorizontal": false,
          "isFlippedVertical": false,
          "isLocked": false,
          "isTemplate": false,
          "isVisible": true,
          "layerListExpandedType": 0,
          "name": "ant-tooltip-arrow",
          "nameIsFixed": false,
          "needsConvertionToNewRoundCorners": false,
          "pointRadiusBehaviour": 1,
          "points": Array [
            Object {
              "_class": "curvePoint",
              "cornerRadius": 0,
              "cornerStyle": 0,
              "curveFrom": "{0, 0}",
              "curveMode": 1,
              "curveTo": "{0, 0}",
              "hasCurveFrom": false,
              "hasCurveTo": false,
              "point": "{0, 0}",
            },
            Object {
              "_class": "curvePoint",
              "cornerRadius": 0,
              "cornerStyle": 0,
              "curveFrom": "{1, 0}",
              "curveMode": 1,
              "curveTo": "{1, 0}",
              "hasCurveFrom": false,
              "hasCurveTo": false,
              "point": "{1, 0}",
            },
            Object {
              "_class": "curvePoint",
              "cornerRadius": 0,
              "cornerStyle": 0,
              "curveFrom": "{1, 1}",
              "curveMode": 1,
              "curveTo": "{1, 1}",
              "hasCurveFrom": false,
              "hasCurveTo": false,
              "point": "{1, 1}",
            },
            Object {
              "_class": "curvePoint",
              "cornerRadius": 0,
              "cornerStyle": 0,
              "curveFrom": "{0, 1}",
              "curveMode": 1,
              "curveTo": "{0, 1}",
              "hasCurveFrom": false,
              "hasCurveTo": false,
              "point": "{0, 1}",
            },
          ],
          "resizingConstraint": 27,
          "resizingType": 0,
          "rotation": 0,
          "shouldBreakMaskChain": false,
          "style": Object {
            "_class": "style",
            "borderOptions": Object {
              "_class": "borderOptions",
              "dashPattern": Array [],
              "isEnabled": true,
              "lineCapStyle": 0,
              "lineJoinStyle": 0,
            },
            "borders": Array [
              Object {
                "_class": "border",
                "color": Object {
                  "_class": "color",
                  "alpha": 0.88,
                  "blue": 0,
                  "green": 0,
                  "red": 0,
                },
                "contextSettings": Object {
                  "_class": "graphicsContextSettings",
                  "blendMode": 0,
                  "opacity": 1,
                },
                "fillType": 0,
                "gradient": Object {
                  "_class": "gradient",
                  "elipseLength": 1,
                  "from": "{0.5, 0}",
                  "gradientType": 0,
                  "stops": Array [],
                  "to": "{0.5, 1}",
                },
                "isEnabled": true,
                "position": 1,
                "thickness": 0,
              },
            ],
            "colorControls": Object {
              "_class": "colorControls",
              "brightness": 0,
              "contrast": 1,
              "hue": 0,
              "isEnabled": false,
              "saturation": 1,
            },
            "contextSettings": Object {
              "_class": "graphicsContextSettings",
              "blendMode": 0,
              "opacity": 1,
            },
            "do_objectID": "UUID",
            "endMarkerType": 0,
            "fills": Array [
              Object {
                "_class": "fill",
                "color": Object {
                  "_class": "color",
                  "alpha": 0.85,
                  "blue": 0,
                  "green": 0,
                  "red": 0,
                },
                "contextSettings": Object {
                  "_class": "graphicsContextSettings",
                  "blendMode": 0,
                  "opacity": 1,
                },
                "fillType": 0,
                "gradient": Object {
                  "_class": "gradient",
                  "elipseLength": 1,
                  "from": "{0.5, 0}",
                  "gradientType": 0,
                  "stops": Array [],
                  "to": "{0.5, 1}",
                },
                "isEnabled": true,
                "noiseIndex": 0,
                "noiseIntensity": 0,
                "patternFillType": 1,
                "patternTileScale": 1,
              },
            ],
            "innerShadows": Array [],
            "miterLimit": 10,
            "shadows": Array [],
            "startMarkerType": 0,
            "windingRule": 1,
          },
        },
      ],
      "name": "编组",
      "nameIsFixed": false,
      "resizingConstraint": 27,
      "resizingType": 0,
      "rotation": -180,
      "shouldBreakMaskChain": false,
      "style": Object {
        "_class": "style",
        "borderOptions": Object {
          "_class": "borderOptions",
          "dashPattern": Array [],
          "isEnabled": true,
          "lineCapStyle": 0,
          "lineJoinStyle": 0,
        },
        "borders": Array [],
        "colorControls": Object {
          "_class": "colorControls",
          "brightness": 0,
          "contrast": 1,
          "hue": 0,
          "isEnabled": false,
          "saturation": 1,
        },
        "contextSettings": Object {
          "_class": "graphicsContextSettings",
          "blendMode": 0,
          "opacity": 1,
        },
        "do_objectID": "UUID",
        "endMarkerType": 0,
        "fills": Array [],
        "innerShadows": Array [],
        "miterLimit": 10,
        "shadows": Array [],
        "startMarkerType": 0,
        "windingRule": 1,
      },
    },
  ],
  "name": "编组",
  "nameIsFixed": false,
  "resizingConstraint": 27,
  "resizingType": 0,
  "rotation": 0,
  "shouldBreakMaskChain": false,
  "style": Object {
    "_class": "style",
    "borderOptions": Object {
      "_class": "borderOptions",
      "dashPattern": Array [],
      "isEnabled": true,
      "lineCapStyle": 0,
      "lineJoinStyle": 0,
    },
    "borders": Array [],
    "colorControls": Object {
      "_class": "colorControls",
      "brightness": 0,
      "contrast": 1,
      "hue": 0,
      "isEnabled": false,
      "saturation": 1,
    },
    "contextSettings": Object {
      "_class": "graphicsContextSettings",
      "blendMode": 0,
      "opacity": 1,
    },
    "do_objectID": "UUID",
    "endMarkerType": 0,
    "fills": Array [],
    "innerShadows": Array [],
    "miterLimit": 10,
    "shadows": Array [],
    "startMarkerType": 0,
    "windingRule": 1,
  },
}
`;
